# Assignment 2 - Individual Checkpoint 1

**Author**: Hanchen Wang

**Group ID**: CC08-3

**Driving problem**: Do they avoid inactivity in at least 10 hours a day?

**Overview of the work**: 
-   By analysing three csv datasets *dailySteps_merged*, *hourlySteps_merged* and *minuteStepsWide_merged*, I will explore the data for participants with id *1503960366*, *1624580081* and *1644430081* and report all aspects required including the number of days of data, daily step count information and minute step count information. 
-   While completing my exploration of the data for three people agreed with my group, Literate Programming is strictly applied to my Jupyter Notebook committed to Github.

**Initial assumptions & predictions**: 
-   The participants in three datasets are consistent. (Has been confirmed in data analysis.)
-   There are independence between different participants in all datasets. (Depends on the data collection process, we assume this is true.)


## Exploratory Data Analysis

### Load Datasets
start_date: 11/9/2024
end_date: 11/9/2024

Three datasets are provided and they are stored in the **src** folder.

We use the **read_csv** function in pandas to read them as dataframe.

In [1]:
# Import all necessary packages
import pandas as pd

# Read three CSV files from the 'src' folder
df_day = pd.read_csv('src/dailySteps_merged.csv')
df_hour = pd.read_csv('src/hourlySteps_merged.csv')
df_min = pd.read_csv('src/minuteStepsWide_merged.csv')

All three datasets have been successfully loaded and we can access them using **df_day**, **df_hour** and **df_min**.
Let's have an overview of those datasets by printing their size and column names.

In [24]:
# Access the size of data set using .shape[0] and .shape[1]
# As .columns return a list of column name, use ',' to join them as a string
print("The size of daily dataset:", df_day.shape[0], "rows and", df_day.shape[1], "columns." )
print("The column names of daily dataset:", ", ".join(df_day.columns))
print()
print("The size of hourly dataset:", df_hour.shape[0], "rows and", df_hour.shape[1], "columns." )
print("The column names of hourly dataset:", ", ".join(df_hour.columns))
print()
print("The size of minute dataset:", df_min.shape[0], "rows and", df_min.shape[1], "columns." )
print("The column names of minute dataset:", ", ".join(df_min.columns))

The size of daily dataset: 940 rows and 3 columns.
The column names of daily dataset: Id, ActivityDay, StepTotal

The size of hourly dataset: 22099 rows and 3 columns.
The column names of hourly dataset: Id, ActivityHour, StepTotal

The size of minute dataset: 21645 rows and 62 columns.
The column names of minute dataset: Id, ActivityHour, Steps00, Steps01, Steps02, Steps03, Steps04, Steps05, Steps06, Steps07, Steps08, Steps09, Steps10, Steps11, Steps12, Steps13, Steps14, Steps15, Steps16, Steps17, Steps18, Steps19, Steps20, Steps21, Steps22, Steps23, Steps24, Steps25, Steps26, Steps27, Steps28, Steps29, Steps30, Steps31, Steps32, Steps33, Steps34, Steps35, Steps36, Steps37, Steps38, Steps39, Steps40, Steps41, Steps42, Steps43, Steps44, Steps45, Steps46, Steps47, Steps48, Steps49, Steps50, Steps51, Steps52, Steps53, Steps54, Steps55, Steps56, Steps57, Steps58, Steps59


All datasets share the **Id** column.

Both daily dataset and hourly dataset have **StepTotal** allowing for easy accessing and direct calculation.

In contrast, the minute dataset is more complex as it has **60 columns** to record step count for one row, which requires extra data manipulation processes.

### Check Missing Values
start_date: 11/9/2024
end_date: 11/9/2024

Before doing any data analysis, we need to make sure all missing values in three datasets are handled properly.

In [2]:
# True will be displayed if any of the column has missing value
df_day.isnull().any()

Id             False
ActivityDay    False
StepTotal      False
dtype: bool

In [3]:
df_hour.isnull().any()

Id              False
ActivityHour    False
StepTotal       False
dtype: bool

In [4]:
df_min.isnull().any()

Id              False
ActivityHour    False
Steps00         False
Steps01         False
Steps02         False
                ...  
Steps55         False
Steps56         False
Steps57         False
Steps58         False
Steps59         False
Length: 62, dtype: bool

Since results for all columns are **False**, there are **no missing values** throughout the three datasets.

### Check Assumption
start_date: 11/9/2024
end_date: 11/9/2024

Firsly, I want to understand the nature of these datasets as well as checking my initial assumption for three datasets.

In [5]:
# Check if unique ids are the same for three datasets
unique_ids_day = df_day['Id'].unique()
unique_ids_hour = df_hour['Id'].unique()
unique_ids_min = df_min['Id'].unique()

print("What are the number of unique ids in daily, hour, minute datasets correspondingly?", 
      len(unique_ids_day), len(unique_ids_hour), len(unique_ids_min))
print("Are unique ids in daily dataset and hour dataset the same?", set(unique_ids_day) == set(unique_ids_hour))
print("Are unique ids in daily dataset and minute dataset the same?", set(unique_ids_day) == set(unique_ids_min))

What are the number of unique ids in daily, hour, minute datasets correspondingly? 33 33 33
Are unique ids in daily dataset and hour dataset the same? True
Are unique ids in daily dataset and minute dataset the same? True


According to the printing message above, values in column **Id** are consistent across three different datasets, which confirms my initial assumption: **The participants in three datasets are consistent.**

### What is the number of days of data for these 3 people?
start_date: 11/9/2024
end_date: 11/9/2024

Now it's time to begin analysing and prepare the report for three selected people.

In [6]:
# Use index 0,1,2 to extract first three value from the list of unique id
id1 = unique_ids_day[0]
id2 = unique_ids_day[1]
id3 = unique_ids_day[2]

# Filter the rows in df_day with corresponding value in 'Id' column using isin()
# Extract 'ActivityDay' column from df_day for id1, id2, id3 using indexing
id1_activity_day = df_day[df_day['Id'].isin([id1])]['ActivityDay']
id2_activity_day = df_day[df_day['Id'].isin([id2])]['ActivityDay']
id3_activity_day = df_day[df_day['Id'].isin([id3])]['ActivityDay']

# Use len() to find out the number of days in the list
# Use unique() to avoid duplicates
id1_num_of_days = len(id1_activity_day.unique())
id2_num_of_days = len(id2_activity_day.unique())
id3_num_of_days = len(id3_activity_day.unique())

# Print out result
print("Ids of three people I'll focus on:", id1, id2, id3)
print("The number of days of data for these 3 people:", id1_num_of_days, id2_num_of_days, id3_num_of_days)

Ids of three people I'll focus on: 1503960366 1624580081 1644430081
The number of days of data for these 3 people: 31 31 30


The result shows a variation in the number of days of data for these 3 people, which could be important to note for further analysis, especially if consistency in the number of days is essential.

### Daily step count information
start_date: 11/9/2024
end_date: 11/9/2024

Let's compute daily average, maximum, minimum, median step count for these 3 people.

We begin with filtering the dataframe using **id1**, **id2** and **id3**.

In [11]:
# Filter the rows in df_day with corresponding value in 'Id' column using isin()
# Extract 'StepTotal' column from df_day for id1, id2, id3 using indexing
id1_daily_step_total = df_day[df_day['Id'].isin([id1])]['StepTotal']
id2_daily_step_total = df_day[df_day['Id'].isin([id2])]['StepTotal']
id3_daily_step_total = df_day[df_day['Id'].isin([id3])]['StepTotal']

# Calculate average step count per day
id1_avg_daily_step_count = int(id1_daily_step_total.mean())
id2_avg_daily_step_count = int(id2_daily_step_total.mean())
id3_avg_daily_step_count = int(id3_daily_step_total.mean())

# Calculate maximum step count
id1_max_daily_step_count = int(id1_daily_step_total.max())
id2_max_daily_step_count = int(id2_daily_step_total.max())
id3_max_daily_step_count = int(id3_daily_step_total.max())

# Calculate minimum step count
id1_min_daily_step_count = int(id1_daily_step_total.min())
id2_min_daily_step_count = int(id2_daily_step_total.min())
id3_min_daily_step_count = int(id3_daily_step_total.min())

# one other observation: Calculate median step count
id1_med_daily_step_count = int(id1_daily_step_total.median())
id2_med_daily_step_count = int(id2_daily_step_total.median())
id3_med_daily_step_count = int(id3_daily_step_total.median())

# Create a dictionary to store the results
daily_result = {
    'ID': [id1, id2, id3],
    'Daily Average Step': [id1_avg_daily_step_count, id2_avg_daily_step_count, id3_avg_daily_step_count],
    'Daily Max Steps': [id1_max_daily_step_count, id2_max_daily_step_count, id3_max_daily_step_count],
    'Daily Min Steps': [id1_min_daily_step_count, id2_min_daily_step_count, id3_min_daily_step_count],
    'Daily Median Steps': [id1_med_daily_step_count, id2_med_daily_step_count, id3_med_daily_step_count]
}

# Create the dataframe for printing as a table
table_day = pd.DataFrame(daily_result)

# Display the table
print(table_day)




           ID  Daily Average Step  Daily Max Steps  Daily Min Steps  \
0  1503960366               12116            18134                0   
1  1624580081                5743            36019             1510   
2  1644430081                7282            18213             1223   

   Daily Median Steps  
0               12207  
1                4026  
2                6683  


Using in-built functions including mean(), max(), min() and median(), all daily step count information needed are extracted and stored in variables with corresponding resonable names. 

Considering the context that all step count data are stored in integers, int() is used to round the calculation result.

From the result shown in the table above, it suggests:
- The user id 1503960366 is consistently active and has a steady daily step count.
- The user id 1624580081 appears to have some intense days of high activity level but otherwise remains much less active on average.
- The user id 1644430081 shows a moderate daily activity level.

### Minute step count information
start_date: 12/9/2024
end_date: 13/9/2024

Due to the higher complexity of the minute dataset, we need to define functions to avoid repetitive code. 

In [8]:
# Function to count number of non-zero minutes
def non_zero_minutes_counter(df):
    # Extract all columns start with 'Steps'
    # Flatten the dataframe into a list
    step_cols = [col for col in df.columns if col.startswith('Steps')]
    
    # Check if value in the step columns is greater than 0 
    # Use first sum() to count non-zero values per row
    # Use second sum() to aggregate across all rows
    non_zero_count = (df[step_cols] > 0).sum().sum()
    
    # Return the total number of non-zero step values as an integer
    return int(non_zero_count)

# Function to count missing value
def missing_minutes_counter(df):
    # Extract & Flatten (Same as above)
    step_cols = [col for col in df.columns if col.startswith('Steps')]

    # Use isnull() to check for missing values
    # Use sum() to count them across all columns
    missing_count = df[step_cols].isnull().sum().sum()

    # Return the total number of missing values as an integer
    return int(missing_count)

# Function to calculate average steps per minute
def avg_steps_per_minute_calculator(df):
    # Extract & Flatten (Same as above)
    step_cols = [col for col in df.columns if col.startswith('Steps')]
    
    # Compute the total sum of steps
    # Divide by the total number of valid (non-missing) entries
    avg_val = df[step_cols].sum().sum() / df[step_cols].count().sum()

    # Return the average as an integer
    return int(avg_val)


# Function to calculate maximum and minimum steps
def max_min_steps_per_minute_calculator(df):
    # Extract & Flatten (Same as above)
    step_cols = [col for col in df.columns if col.startswith('Steps')]

    # Use first max() used to find maximum in each column
    # Use second max() used to find maximum of the maximums
    max_val = df[step_cols].max().max()

    # Use first min() to find minimum in each column
    # Use second min() to find minimum of the minimums
    min_val = df[step_cols].min().min()

    # Return both the maximum and minimum as integers
    return int(max_val), int(min_val)


# one other observation: Function to calculate standard deviation of step count
def sd_steps_per_minute_calculator(df):
    # Extract & Flatten (Same as above)
    step_cols = [col for col in df.columns if col.startswith('Steps')]
    
    # Use stack() to flatten all columns into one 
    # Use std() to calculate the standard deviation
    sd = df[step_cols].stack().std()

    # Return the standard deviation as an integer
    return int(sd)


All the functions needed have been defined above.

We then need to filter out the df_min for id1, id2 and id3 individually for the input of those functions.

Finally, apply the functions and save the results into variables with reasonable names.


In [13]:
# Filter the rows in df_min with corresponding value in 'Id' column using isin()
id1_minutes = df_min[df_min['Id'].isin([id1])]
id2_minutes = df_min[df_min['Id'].isin([id2])]
id3_minutes = df_min[df_min['Id'].isin([id3])]


# Apply all functions above for three people using filtered dataframes
id1_non_zero_minutes = non_zero_minutes_counter(id1_minutes)
id2_non_zero_minutes = non_zero_minutes_counter(id2_minutes)
id3_non_zero_minutes = non_zero_minutes_counter(id3_minutes)

id1_missing_minute = missing_minutes_counter(id1_minutes)
id2_missing_minute = missing_minutes_counter(id2_minutes)
id3_missing_minute = missing_minutes_counter(id3_minutes)

id1_avg_step_count = avg_steps_per_minute_calculator(id1_minutes)
id2_avg_step_count = avg_steps_per_minute_calculator(id2_minutes)
id3_avg_step_count = avg_steps_per_minute_calculator(id3_minutes)

id1_max_minute_step_count, id1_min_minute_step_count = max_min_steps_per_minute_calculator(id1_minutes)
id2_max_minute_step_count, id2_min_minute_step_count = max_min_steps_per_minute_calculator(id2_minutes)
id3_max_minute_step_count, id3_min_minute_step_count = max_min_steps_per_minute_calculator(id3_minutes)

id1_sd_minute_step_count = sd_steps_per_minute_calculator(id1_minutes)
id2_sd_minute_step_count = sd_steps_per_minute_calculator(id2_minutes)
id3_sd_minute_step_count = sd_steps_per_minute_calculator(id3_minutes)


# Create a dictionary to store the results
minute_result = {
    'ID': [id1, id2, id3],
    'Non-Zero Minutes': [id1_non_zero_minutes, id2_non_zero_minutes, id3_non_zero_minutes],
    'Missing Minutes': [id1_missing_minute, id2_missing_minute, id3_missing_minute],
    'Minute Average Steps': [id1_avg_step_count, id2_avg_step_count, id3_avg_step_count],
    'Minute Max Steps': [id1_max_minute_step_count, id2_max_minute_step_count, id3_max_minute_step_count],
    'Minute Min Steps': [id1_min_minute_step_count, id2_min_minute_step_count, id3_min_minute_step_count],
    'Minute Standard Deviation of Steps': [id1_sd_minute_step_count, id2_sd_minute_step_count, id3_sd_minute_step_count]
}

# Create the dataframe for printing as a table
table_minute = pd.DataFrame(minute_result)

# Display the table
print(table_minute)



           ID  Non-Zero Minutes  Missing Minutes  Minute Average Steps  \
0  1503960366              8311                0                     8   
1  1624580081              3679                0                     3   
2  1644430081              5978                0                     5   

   Minute Max Steps  Minute Min Steps  Minute Standard Deviation of Steps  
0               165                 0                                  25  
1               184                 0                                  17  
2               134                 0                                  17  



According to the result table of minute step count information above, we can see some patterns:
- The user id 1503960366 is the most active among three people due to the highest non-zero minutes of 8311 and average steps per minute of 8. This user's  standard deviation of step count of 25 indicates higher variability in acitivity levels than other two users.
- Although user id 1624580081 has the maximum step count in one minute of 184, this person is still the least active user due to its lowest non-zero minutes of 3679 and average step count of 3.

## Conclusion
start_date: 13/9/2024
end_date: 13/9/2024

In this checkpoint 1, we performed an initial exploration of the step count data focusing on three people. After calculating all information needed, they are saved into two dataframes **table_day** and **table_minute**, available for further analysis in next checkpoint such as data visualization.

As a final statement, I learnt there is no missing values throughout three datasets, while variations in the number of days of data recorded for different participants and their activity level appear to be obvious. This highly relates to my driving problem "Do they avoid inactivity in at least 10 hours a day?":

1. Since there are no missing values, I can confidently analyze the step data without concerns about incomplete records, ensuring a realiable analysis result.
2. The varying number of days recorded for different participants suggests that some individuals have more or less data available for analysis, which can affect the accuracy of the conclusions of their activity pattern over time.
3. Based on an initial data exploration for three people, I discovered that while some participants are active for more than 10 hours a day, some may struggle to maintain a consistent active level. Such variations deserve more data anlysis for understanding the activity level in this group of participants and producing meaningful results.
